blocks/components/editable/index.js

```diff
@@ -223,66 +216,142 @@ export default class Editable extends wp.element.Component {
-		// These calls can make the focus jump
+	getChildIndex( child ) {
+		return Array.from( child.parentNode.childNodes ).indexOf( child );
+	}
+
+	fillPath( path, node, rootNode ) {
+		while ( node !== rootNode ) {
+			path.unshift( this.getChildIndex( node ) );
+			node = node.parentNode;
+		}
 
-		this.editor.save();
+		return path;
 	}
 
-	setContent( content ) {
-		if ( ! content ) {
-			content = '';
+	getSelection() {
+		const {
+			startOffset,
+			startContainer,
+			endOffset,
+			endContainer,
+			collapsed
+		} = this.editor.selection.getRng();
+		const rootNode = this.editor.getBody();
+
+		return {
+			start: this.fillPath( [ startOffset ], startContainer, rootNode ),
+			end: this.fillPath( [ endOffset ], endContainer, rootNode ),
+			isCollapsed: collapsed
+		};
+	}
+
+	findNodeWithPath( path, rootNode ) {
+		const index = path[ 0 ];
+
+		if ( index === undefined || ! rootNode.hasChildNodes() ) {
+			return;
 		}
 
-		content = wp.element.renderToString( content );
-		this.editor.setContent( content, { format: 'raw' } );
+		const node = rootNode.childNodes[ index ];
+
+		if ( ! node ) {
+			return;
+		}
+
+		const newPath = drop( path );
+
+		if ( node.nodeType === 3 ) {
+			return { node, index: newPath[ 0 ] || 0 };
+		}
+
+		if ( newPath.length ) {
+			return this.findNodeWithPath( newPath, node );
+		}
+
+		return {
+			node: node.parentNode,
+			index: this.getChildIndex( node )
+		};
+	}
+
+	setSelection( { start, end } ) {
+		if ( ! start || ! start.length ) {
+			return;
+		}
+
+		const rootNode = this.editor.getBody();
+		const range = this.editor.dom.createRng();
+		const currentRange = this.editor.selection.getRng();
+		const startSet = this.findNodeWithPath( start, rootNode );
+		const endSet = this.findNodeWithPath( end, rootNode );
+		const propsToCompare = [
+			'startOffset', 'endOffset',
+			'startContainer', 'endContainer'
+		];
+
+		if ( ! startSet || ! endSet ) {
+			return;
+		}
+
+		range.setStart( startSet.node, startSet.index );
+		range.setEnd( endSet.node, endSet.index );
+
+		if ( ! isEqual( pick( currentRange, propsToCompare ), pick( range, propsToCompare ) ) ) {
+			this.editor.selection.lastFocusBookmark = null;
+			this.editor.selection.setRng( range );
+		} else if ( document.activeElement !== rootNode ) {
+			this.editor.focus();
+		}
+	}
+
+	setContent( content ) {
+		this.content = content || [];
+		this.editor.setContent( wp.element.renderToString( this.content ) );
 	}
 
 	getContent() {
 		return nodeListToReact( this.editor.getBody().childNodes || [], createElement );
 	}
 
-	updateFocus() {
+	maybeFocus() {
 		const { focus } = this.props;
+
 		if ( focus ) {
 			this.editor.focus();
 			// Offset = -1 means we should focus the end of the editable
 			if ( focus.offset === -1 ) {
 				this.editor.selection.select( this.editor.getBody(), true );
 				this.editor.selection.collapse( false );
+				this.selection = this.getSelection();
```

Comment:
Is this needed? I mean, I'd think these statements would trigger a `onSelectionChange`?